File: exttool/Joypad.cs
using System.Collections.Generic;

namespace BirdsEye {
    public class Joypad {
        public string? Name;
        public IDictionary<string, bool>? Controls;
        public IDictionary<string, int?>? ControlsAnalog;
        public int? DefaultController;
    }

    public class NESJoypad : Joypad {
        public NESJoypad() {
            Name = "NES";
            Controls = new Dictionary<string, bool>() {
                {"A", false}, {"B", false}, {"Up", false}, {"Down", false}, {"Right", false},
                {"Left", false}, {"Select", false}, {"Start", false},
            };
            DefaultController = 1;
        }
    }

    public class GBAndGBCJoypad : Joypad {
        public GBAndGBCJoypad() {
            Name = "GB(C)";
            Controls = new Dictionary<string, bool>() {
                {"A", false}, {"B", false}, {"Up", false}, {"Down", false}, {"Right", false},
                {"Left", false}, {"Select", false}, {"Start", false},
            };
            DefaultController = 1;
        }
    }

    public class SNESJoypad : Joypad {
        public SNESJoypad() {
            Name = "SNES";
            Controls = new Dictionary<string, bool>() {
                {"A", false}, {"B", false}, {"L", false}, {"R", false}, {"X", false}, {"Y", false},
                {"Up", false}, {"Down", false}, {"Right", false}, {"Left", false},
                {"Select", false}, {"Start", false},
            };
            DefaultController = 1;
        }
    }

    public class NDSJoypad : Joypad {
        public NDSJoypad() {
            Name = "NDS";
            Controls = new Dictionary<string, bool>() {
                {"A", false}, {"B", false}, {"L", false}, {"R", false}, {"X", false}, {"Y", false},
                {"Up", false}, {"Down", false}, {"Right", false}, {"Left", false},
                {"Select", false}, {"Start", false}, {"LidClose", false}, {"LidOpen", false},
                {"Touch", false},
            };
            ControlsAnalog = new Dictionary<string, int?>() {
                {"Mic Volume", 0}, {"GBA Light Sensor", 0}, {"Touch X", 142}, {"Touch Y", 0},
            };
            DefaultController = null;
        }
    }
}

File: birdseyelib/joypad.py
class Joypad:
    """Treated as a custom type for joypad layouts."""
    pass


class NESJoypad(Joypad):
    """Controller mapping for the NES.
    
    The \"Power\" and \"Reset\" buttons are not implemented for the time being."""
    def __init__(self):
        super().__init__()
        self._name = "NES"
        self.controls = {
            "A": False, "B": False, "Up": False, "Down": False, "Right": False, "Left": False,
            "Select": False, "Start": False,
        }


class GBAndGBCJoypad(Joypad):
    """Controller mapping for the GB and GBC.
    
    The \"Power\" button is not implemented for the time being."""
    def __init__(self):
        super().__init__()
        self._name = "GB(C)"
        self.controls = {
            "A": False, "B": False, "Up": False, "Down": False, "Right": False, "Left": False,
            "Select": False, "Start": False,
        }


class SNESJoypad(Joypad):
    """Controller mapping for the SNES.
    
    The \"Power\" and \"Reset\" buttons are not implemented for the time being."""
    def __init__(self):
        super().__init__()
        self._name = "SNES"
        self.controls = {
            "A": False, "B": False, "L": False, "R": False, "X": False, "Y": False, "Up": False,
            "Down": False, "Right": False, "Left": False, "Select": False, "Start": False,
        }


class NDSJoypad(Joypad):
    """Controller mapping for the NDS
    
    The \"Power\" button is not implemented for the time being."""
    def __init__(self):
        super().__init__()
        self._name = "NDS"
        self.controls = {
            "A": False, "B": False, "L": False, "R": False, "X": False, "Y": False, "Up": False,
            "Down": False, "Right": False, "Left": False, "Select": False, "Start": False,
            "LidClose": False, "LidOpen": False, "Touch": False,
        }
        self.analog_controls = {
            "Mic Volume": "0", "GBA Light Sensor": "0",  "Touch X": "142", "Touch Y": "0",
        }
